extension/jmxmetricsextension/subprocess/subprocess.go

```diff
@@ -0,0 +1,282 @@
+		config:         conf,
+		logger:         logger,
+		shutdownSignal: make(chan struct{}),
+		sendToStdIn:    sendToStdIn,
+	}
+}
+
+// A global var that is available only for testing
+var restartDelay = 5 * time.Second
+
+const (
+	Starting     = "Starting"
+	Running      = "Running"
+	ShuttingDown = "ShuttingDown"
+	Stopped      = "Stopped"
+	Restarting   = "Restarting"
+	Errored      = "Errored"
+)
+
+func (subprocess *Subprocess) Start(ctx context.Context) error {
+	var cancelCtx context.Context
+	cancelCtx, subprocess.cancel = context.WithCancel(ctx)
+
+	for k, v := range subprocess.config.EnvironmentVariables {
+		joined := fmt.Sprintf("%v=%v", k, v)
+		subprocess.envVars = append(subprocess.envVars, joined)
+	}
+
+	go func() {
+		subprocess.run(cancelCtx) // will block for lifetime of process
+		close(subprocess.shutdownSignal)
+	}()
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (subprocess *Subprocess) Shutdown(ctx context.Context) error {
+	subprocess.cancel()
+	t := time.NewTimer(5 * time.Second)
+
+	// Wait for the subprocess to exit or the timeout period to elapse
+	select {
+	case <-ctx.Done():
+	case <-subprocess.shutdownSignal:
+	case <-t.C:
+	}
+
+	return nil
+}
+
+// Core event loop
+func (subprocess *Subprocess) run(ctx context.Context) {
+	var cmd *exec.Cmd
+	var err error
+	var stdin io.WriteCloser
+	var stdout io.ReadCloser
+
+	// writer is signalWhenProcessReturned() and closer is this loop, so we need synchronization
+	processReturned := make(chan error)
+	processReturnedOpen := &atomic.Value{}
+	processReturnedOpen.Store(true)
+	processReturnedLock := &sync.Mutex{}
+
+	state := Starting
+	for {
+		subprocess.logger.Debug("subprocess changed state", zap.String("state", state))
+
+		switch state {
+		case Starting:
+			cmd, stdin, stdout = createCommand(
+				subprocess.config.ExecutablePath,
+				subprocess.config.Args,
+				subprocess.envVars,
+			)
+
+			go collectStdout(stdout, subprocess.Stdout, subprocess.logger)
+
+			subprocess.logger.Debug("starting subprocess", zap.String("command", cmd.String()))
+			err = cmd.Start()
+			if err != nil {
+				state = Errored
+				continue
+			}
+			subprocess.pid.setPid(cmd.Process.Pid)
+
+			go signalWhenProcessReturned(cmd, processReturned, processReturnedLock, processReturnedOpen)
+
+			state = Running
+		case Running:
+			err = subprocess.sendToStdIn(subprocess.config.StdInContents, stdin)
+			stdin.Close()
+			if err != nil {
+				state = Errored
+				continue
+			}
+
+			select {
+			case err = <-processReturned:
+				if err != nil && ctx.Err() == nil {
+					err = fmt.Errorf("unexpected shutdown: %w", err)
+					// We aren't supposed to shutdown yet so this is an error state.
+					state = Errored
+					continue
+				}
+				// We must close this channel or can wait indefinitely at ShuttingDown
+				processReturnedLock.Lock()
+				close(processReturned)
+				processReturnedOpen.Store(false)
+				processReturnedLock.Unlock()
```

Comment:
This probably could be moved into a method once you have a struct to hold these values.

Reply:
Broke out into a helper, which seems a bit more clear.